src/branch_and_bound.rs

```diff
@@ -0,0 +1,582 @@
+
+    if available_value < target {
+        return Some(Vec::new());
+    }
+
+    utxo_candidate_amounts.sort();
+    utxo_candidate_amounts.reverse();
+
+    while iteration < ITERATION_LIMIT {
+        // There are two conditions for backtracking:
+        //
+        // 1_ Not enough value to make it to target.
+        //    This condition happens before reaching a leaf node.
+        //    Looking for a leaf node condition should not make a difference.
+        //    This backtrack removes more than one node and instead starts
+        //    the exploration of a new subtree.
+        //
+        // From:
+        //      o
+        //     / \
+        //    4
+        //   / \
+        //      3
+        //     / \
+        //        2
+        //       /
+        //      1
+        //
+        // To:
+        //      o
+        //     / \
+        //        4
+        //       / \
+        //      3
+        //
+        //
+        // 2_ value meets or exceeded target.
+        //    In this condition, we only backtrack one node
+        //
+        // From:
+        //      o
+        //     /
+        //    4
+        //   /
+        //  3
+        //
+        // To:
+        //      o
+        //     /
+        //    4
+        //   / \
+        //      3
+
+        // Set initial loop state
+        backtrack = false;
+        backtrack_subtree = false;
+
+        // * not enough value to make it to the target.
+        //   Therefore, explore a new new subtree.
+        if available_value + value < target {
+            backtrack_subtree = true;
+        }
+        // This optimization provides an upper bound on the amount of waste that is acceptable.
+        // Since value is lost when we create a change output due to increasing the size of the
+        // transaction by an output, we accept solutions that may be larger than the target as
+        // if they are exactly equal to the target and consider the overage waste or a throw
+        // away amount.  However we do not consider values greater than value + cost_of_change.
+        //
+        // This effectively creates a range of possible solution where;
+        // range = (target, target + cost_of_change]
+        //
+        // That is, the range includes solutions that exactly equal the target up to but not
+        // including values greater than target + cost_of_change.
+        else if value > target + cost_of_change {
+            backtrack = true;
+        }
+        // * value meets or exceeds the target.
+        //   Record the solution and the waste then continue.
+        //
+        // Check if index_selection is better than the previous known best, and
+        // update best_selection accordingly.
+        else if value >= target {
+            backtrack = true;
+
+            let current_waste = value - target;
+
+            if current_waste <= waste {
+                best_selection = Some(index_selection.clone());
```

Comment:
```rust
match best_selection {
    None => best_selection = Some(index_selection.clone());
    Some(best_selection) => {
        best_selection.clear();
        best_selection.extend_from_slice(&index_selection);
    },
}
```

Reply:
interestingly this seems to have worsened performance slightly.  Note here is what made the compiler happy:

```
            if current_waste <= waste {
                match best_selection {
                    None => best_selection = Some(index_selection.clone()),
                    Some(ref mut best_selection) => {
                        best_selection.clear();
                        best_selection.extend_from_slice(&index_selection);
                    }
                }

                waste = current_waste;
            }

```

I don't think avoiding the reassignment of the stack variable `best_selection` is better here, and based on the benchmarks actually a bit worse.

from about ~830,000 to ~850,000 using the exiting benchmark.

Reply:
s/exiting/existing

Reply:
That's super weird. I know it's not idiomatic but doesn't empty vec work in place of `None`? Maybe the branch is slowing it down.

Another crazy idea: maybe pre-allocate the vectors to the size of the inputs. It wastes quite a chunk of memory but could be much faster and we could trim it at the end (or let the user trim it).

Reply:
Yeah I thought of that too, and maybe it's worth doing.  It's possible it doesn't waste any memory of the selection happens to use all inputs but it's unlikely.

Reply:
The thing is that it doesn't seem like a fair comparison anymore between the Rust and C++ implementation to pre-allocate and consume more memory for the sake of performance.  I would think the same option would be available in the core C++ implementation and perhaps it's reasoned that the additional memory usage isn't worth the perf increase.

Reply:
> That's super weird. I know it's not idiomatic but doesn't empty vec work in place of None

As commented in the `/// Returns sections` `None` means there was some type of error such as overflow while empty vec means that everything ran ok, but no match was found.  I'm flexible on the semantics but I think that makes sense as downstream components may want to handle an error different than no results.

Reply:
You can still return `Option` but internally use empty array and only reurn `None` on errors.